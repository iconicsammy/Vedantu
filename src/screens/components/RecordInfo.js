/* eslint-disable react/jsx-closing-bracket-location */
import PropTypes from 'prop-types';
import React from 'react';
import {View, Text, TouchableOpacity, Alert} from 'react-native';
import style from 'screens/style.js';
import Icon from 'react-native-vector-icons/Ionicons';

const RecordInfo = (props) => {
  const {recordDetail, onPress} = props;

  const showPhone = (phone) => {
    Alert.alert('Phone Number Is', phone);
  };

  return (
    <>
      <TouchableOpacity onPress={onPress}>
        <View style={style.recordContainer}>
          <View style={style.nameInitialContainer}>
            <Text style={style.nameInitial}>{recordDetail.name.charAt(0)}</Text>
          </View>
          <View style={style.detailContainer}>
            <Text style={style.name}>{recordDetail.name}</Text>
            <Text>{recordDetail.categoryName}</Text>
            <Text>{recordDetail.location}</Text>
            <Text>{recordDetail.providerType}</Text>
          </View>
          <View style={style.status}>
            <Text style={style.timeAgo}>16 hours ago</Text>
            <Icon
              name="phone-portrait-sharp"
              size={30}
              color="#900"
              style={style.icon}
              onPress={() => showPhone(recordDetail.phoneNumber)}
            />
          </View>
        </View>
      </TouchableOpacity>
    </>
  );
};

RecordInfo.propTypes = {
  onPress: PropTypes.func,
  recordDetail: PropTypes.shape({
    name: PropTypes.string.isRequired,
    categoryName: PropTypes.string.isRequired,
    location: PropTypes.string.isRequired,
    providerType: PropTypes.string.isRequired,
    phoneNumber: PropTypes.string.isRequired,
  }).isRequired,
};

RecordInfo.defaultProps = {
  onPress: undefined,
};

export default RecordInfo;
