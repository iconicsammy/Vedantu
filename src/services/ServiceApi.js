import axios from 'axios';
import APIS from '../../src/APIs';

class ApiService {
  constructor() {
    if (!ApiService.instance) {
      ApiService.instance = this;
    }
    // Initialize object
    return ApiService.instance;
  }

  static getInstance() {
    if (!ApiService.myInstance) {
      ApiService.myInstance = new ApiService();
    }
    return this.myInstance;
  }

  getData = async () => {
    try {
      const res = await axios.get(APIS.mocky_io, {});
      return res.data.dataList;
    } catch (error) {
      return error;
    }
  };
}
const ServiceCApi = new ApiService();

export default ServiceCApi;
