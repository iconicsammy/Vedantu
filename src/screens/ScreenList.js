import PropTypes from 'prop-types';
import React, {useEffect, useState, useCallback} from 'react';
import {View, ScrollView, Text, RefreshControl} from 'react-native';
import ServiceApi from 'services/ServiceApi';
import Spinner from 'screens/components/Spinner';
import {saveListAction, setActiveRecordAction} from 'store/actions/data.action';
import RecordInfo from 'screens/components/RecordInfo';
import {useDispatch} from 'react-redux';
import {useSelector} from 'react-redux';
import style from 'screens/style.js';

const ScreenList = ({navigation}) => {
  const records = useSelector((state) => state.data);
  const dispatch = useDispatch();
  const [refreshing, setRefreshing] = useState(false);

  const getList = async () => {
    /*
      get list from the web
    */

    setRefreshing(true);

    try {
      const result = await ServiceApi.getData();
      dispatch(saveListAction(result));
      setRefreshing(false);
    } catch (error) {
      setRefreshing(false);
    }
  };

  const onPullRefresh = useCallback(async () => {
    setRefreshing(true);
    await getList();
    setRefreshing(false);
  }, []);

  useEffect(() => {
    getList();
  }, []);

  const openDetail = (record) => {
    dispatch(setActiveRecordAction(record));
    navigation.navigate('Detail');
  };

  return (
    <>
      {refreshing === false ? (
        <>
          <ScrollView
            refreshControl={
              <RefreshControl
                refreshing={refreshing}
                onRefresh={onPullRefresh}
              />
            }>
            <View style={style.mainContainer}>
              {records.map((record) => (
                <RecordInfo
                  onPress={() => openDetail(record)}
                  key={record.enqId.toString()}
                  recordDetail={record}
                />
              ))}
            </View>
          </ScrollView>
        </>
      ) : (
        <Spinner />
      )}
    </>
  );
};
ScreenList.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }).isRequired,
};
export default ScreenList;
