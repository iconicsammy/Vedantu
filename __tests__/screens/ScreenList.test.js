/* eslint-disable prefer-destructuring */
/* eslint-disable jest/valid-expect */
import React from 'react';
import {
  render,
  act,
  flushMicrotasksQueue,
  fireEvent,
} from 'react-native-testing-library';
import {mockNavigationProps} from '../testhelpers';
import RecordInfo from 'screens/components/RecordInfo';
import * as ReactRedux from 'react-redux';
import {RefreshControl, View} from 'react-native';
import ServiceApi from 'services/ServiceApi';
import * as mockData from '../mockData.json';
import ScreenList from 'screens/ScreenList';

describe('actvity ScreenList', () => {
  let spyServiceApiGetData;
  let spyUseSelector;
  const props = mockNavigationProps({}, {});

  beforeEach(() => {
    spyServiceApiGetData = jest.spyOn(ServiceApi, 'getData');
    spyUseSelector = jest.spyOn(ReactRedux, 'useSelector');
  });

  afterEach(() => {
    spyServiceApiGetData.mockReset();
    spyUseSelector.mockReset();
  });

  it('renders correctly', async () => {
    spyUseSelector.mockReturnValue(mockData.dataList);
    spyServiceApiGetData.mockResolvedValue([]);

    const snapShotTree = render(
      <ScreenList navigation={props.navigation} />,
    ).toJSON();

    await act(async () => {
      await flushMicrotasksQueue();
    });
    expect(snapShotTree).toMatchSnapshot();
  });

  it('should render all data on the screen', async () => {
    spyServiceApiGetData.mockResolvedValue(mockData.dataList);
    spyUseSelector.mockReturnValue(mockData.dataList);
    const {UNSAFE_getAllByType} = render(
      <ScreenList navigation={props.navigation} />,
    );

    await act(async () => {
      await flushMicrotasksQueue();
    });
    const dataRecordViews = UNSAFE_getAllByType(RecordInfo);
    expect(dataRecordViews).toHaveLength(mockData.dataList.length);
    expect(spyServiceApiGetData).toHaveBeenCalled();
    // click all elements to make sure we open their dedicated screen
    dataRecordViews.forEach((g, index) => {
      fireEvent.press(g);
      expect(props.navigation.navigate).toHaveBeenCalledWith('Detail');
    });
  });

  it('should render groups again on pull refresh', async () => {
    spyServiceApiGetData.mockResolvedValue(mockData.dataList);
    spyUseSelector.mockReturnValue(mockData.dataList);
    const wrapper = render(<ScreenList navigation={props.navigation} />);

    await act(async () => {
      await flushMicrotasksQueue();
    });

    const refreshControl = wrapper.UNSAFE_getByType(RefreshControl);
    // pull refresh
    fireEvent(refreshControl, 'onRefresh', {});
    await act(async () => {
      await flushMicrotasksQueue();
    });
    const dataRecordViews = wrapper.UNSAFE_getAllByType(RecordInfo);
    expect(dataRecordViews).toHaveLength(mockData.dataList.length);
    expect(spyServiceApiGetData).toHaveBeenCalled();
    // click all elements to make sure we open their dedicated screen
    dataRecordViews.forEach((g, index) => {
      fireEvent.press(g);
      expect(props.navigation.navigate).toHaveBeenCalledWith('Detail');
    });
  });

});
