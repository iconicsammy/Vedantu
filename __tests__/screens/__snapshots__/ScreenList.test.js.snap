// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actvity ScreenList renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": 10,
      },
    ]
  }
>
  <ActivityIndicator
    animating={true}
    color="#148774"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 280,
        "justifyContent": "center",
      }
    }
  />
</View>
`;
